Add sitemap.xml

waxlamp opened the pull request:
This PR adds a sitemap.xml view to the DANDI backend that should help with Google presenting search results for individual Dandisets. Together with activating prerendering through Netlify (see #2663), this should solve most of our (current) SEO problems.

### Details
- The sitemap.xml view is generated by Django's sitemaps framework
- A new robots.txt view presents a `Sitemap` record to visiting crawlers
- Both of these views are generated serverside and proxied to the frontend (using vite server proxy config for dev, and Netlify redirects for the sandbox and production deployments)
- While the sitemap is always generated, there's a new setting to control whether the sitemap *entries* are generated. This is because we will not want all DANDI deployments to be search engine indexed. The most salient example is the sandbox deployment: we only want Google to turn up references to production, so we'll leave this turned off for sandbox. This setting also controls whether the `Sitemap:` directive appears in the robots.txt view.

### Deployment
When this PR is merged/released, it will not have any immediate effect since the feature defaults to inactive. To activate it on select deployments, we need to set the `DJANGO_DANDI_ENABLE_SITEMAP_XML` environment variable to `True`; to do so for our prod deployment, the best way would be to send a PR to the `dandi-infrastructure` repo.

### TODO
- [ ] after merge/release, set up a dandi-infrastructure PR to activate for prod
- [ ] after merge/release, submit site to Google for reindexing

Closes #752 (this was already closed, but the current PR addresses it)
Closes #2230

waxlamp committed:
Add sitemap for DLPs

waxlamp committed:
Use a setting to inject the domain into the sitemap.xml content

waxlamp committed:
Add sitemap to robots.txt

waxlamp committed:
Add setting to toggle sitemap.xml generation

We don't want every DANDI deployment indexable by search engines. For
example, we will want to activate this for the prod deployment, but
leave it off for sandbox.

waxlamp reviewed (commented):
I think this is pretty straightforward, so I'm just looking for obvious pitfalls.

waxlamp commented on dandiapi/urls.py:
```diff
@@ -163,6 +165,16 @@ def to_url(self, value):
+            {'sitemaps': sitemaps},
+            name='django.contrib.sitemaps.views.sitemap',
+        ),
+    ]
```
Forgot to mention it in my self-review, but is there any need to cache this view?

My guess is there's no need, at least not at the scale we're currently running at. Apparently, past 50000 entries, you have to split your sitemap.xml file up or risk Google ignoring you, but of course we're nowhere near that number.

jjnesbitt replied:
I don't see a need, not at the moment anyway.

jjnesbitt commented on dandiapi/sitemaps.py:
```diff
@@ -0,0 +1,25 @@
+        return f'/dandiset/{obj.identifier}/'
+
+    def get_domain(self, site):
+        return settings.DANDI_WEB_APP_DOMAIN
```
Since this is only used in one place, maybe we can avoid adding a new setting:
```suggestion
        return urlparse(settings.DANDI_WEB_APP_URL).netloc
```

waxlamp replied:
Good idea; see 25352f5b2dc6f8d7bf36b14a4fd6d65e7a1c3290.

jjnesbitt reviewed (changes requested):
Looks good overall. My only lingering concern is that we're generating sitemaps for a different domain than the one the API is hosted at (`api.dandiarchive.org` vs `dandiarchive.org`). Is there any issue with this?

waxlamp committed:
Inline the "domain" setting to its single point of use

Co-authored-by: Jacob Nesbitt <jjnesbitt2@gmail.com>

waxlamp committed:
Move the sitemaps view to a non-canonical URL

This is to prevent crawlers from reading the sitemap directly from the
backend, since it references URLs not on the same host, which is not
allowed (see https://www.sitemaps.org/protocol.html#location).

This is in preparation for generating the sitemap server-side, but
proxying it to the frontend so that the hosts align properly.

waxlamp committed:
Proxy the sitemap view to the frontend in development setup

waxlamp committed:
Proxy sitemap view to Netlify deployments

This copies the pattern established by the "server-info-build" plugin.

waxlamp committed:
Change how DANDI_ENABLE_SITEMAP_XML works

With this change, the backend always generates a sitemap file, but the
ENABLE setting governs whether any URL entries are included in it. This
simplifies deployment concerns: the backend always generates a sitemap,
and the frontend always proxies it to `/sitemap.xml`, but now a
deployment not wishing to advertise DLPs will simply not show any of
those URLs.

waxlamp committed:
Generate robots.txt serverside (and proxy it to the frontend)

This commit restores the original robots.txt view for the backend, and
creates a new one specifically for the frontend. It's done this way to
make it easier to generate the frontend's sitemap.xml URL within
robots.txt (and it piggybacks off the similar proxying trick for
sitemap.xml itself).

waxlamp commented on web/.env.development:
```diff
@@ -1,3 +1,4 @@
+VITE_APP_DANDI_BACKEND_ROOT=http://localhost:8000
 VITE_APP_DANDI_API_ROOT=http://localhost:8000/api/
 VITE_APP_OAUTH_API_ROOT=http://localhost:8000/oauth/
```
I think these can be refactored now to inject the `DANDI` and `API` variables into the config at runtime, but it seemed like it was out of scope for this PR.

waxlamp commented:
@jjnesbitt, this is ready for another review. There are some choices in the PR now that may seem a bit odd, so I'm happy to discuss any time.

waxlamp committed:
Use a simpler view name

Co-authored-by: Jacob Nesbitt <jjnesbitt2@gmail.com>

waxlamp committed:
Don't even install the sitemap view when disabled

This is to be totally explicit with, e.g., Googlebot that when we're not
serving a sitemap file, there isn't even an empty file to present. This
matters only for minor reasons, such as avoiding warnings from Google
Search Console about an empty sitemap, etc. Or, put another way, it
eliminates any possible issues that might crop up when presenting an
empty sitemap file to search engines (as opposed to not presenting any
file at all).
